File: eel-core/src/main/kotlin/io/eels/component/hive/dialect/ParquetHiveDialect.scala
package io.eels.component.hive.dialect

import com.typesafe.scalalogging.slf4j.StrictLogging
import io.eels.component.avro.{AvroSchemaFn, ConvertingAvroRecordMarshaller}
import io.eels.component.hive.{HiveDialect, HiveWriter, Predicate}
import io.eels.component.parquet.{ParquetIterator, ParquetLogMute, ParquetReaderSupport, RollingParquetWriter}
import io.eels.{InternalRow, Schema, SourceReader}
import org.apache.hadoop.fs.{FileSystem, Path}

object ParquetHiveDialect extends HiveDialect with StrictLogging {

  override def writer(schema: Schema, path: Path)
                     (implicit fs: FileSystem): HiveWriter = {
    ParquetLogMute()

    // hive is case insensitive so we must lower case everything to keep it consistent
    lazy val avroSchema = AvroSchemaFn.toAvro(schema, caseSensitive = false)
    lazy val writer = RollingParquetWriter(path, avroSchema)
    lazy val marshaller = new ConvertingAvroRecordMarshaller(avroSchema)
    var count = 0l

    new HiveWriter {
      override def close(): Unit = if (count > 0) writer.close()
      override def write(row: InternalRow): Unit = {
        val record = marshaller.toRecord(row)
        writer.write(record)
        count = count + 1
      }
    }
  }

  override def reader(path: Path, dataSchema: Schema, targetSchema: Schema, predicate: Option[Predicate])
                     (implicit fs: FileSystem): SourceReader = new SourceReader {
    require(targetSchema.columns.nonEmpty, "Cannot create parquet reader for a target schema that has no columns")
    val reader = ParquetReaderSupport.createReader(path, targetSchema.size < dataSchema.size, predicate, targetSchema)
    override def close(): Unit = reader.close()
    override def iterator: Iterator[InternalRow] = ParquetIterator(reader, targetSchema)
  }
}

File: eel-core/src/main/kotlin/io/eels/component/hive/dialect/ParquetHiveDialect.kt
package io.eels.component.hive.dialect

import io.eels.Row
import io.eels.component.Predicate
import io.eels.component.avro.AvroRecordSerializer
import io.eels.component.avro.AvroSchemaFns
import io.eels.component.hive.HiveDialect
import io.eels.component.hive.HiveWriter
import io.eels.component.parquet.ParquetRowIterator
import io.eels.component.parquet.ParquetLogMute
import io.eels.component.parquet.ParquetReaderFns
import io.eels.component.parquet.ParquetRowWriter
import io.eels.schema.Schema
import io.eels.util.Logging
import io.eels.util.Option
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.Path
import rx.Observable

object ParquetHiveDialect : HiveDialect, Logging {

  override fun read(path: Path,
                    metastoreSchema: Schema,
                    projectionSchema: Schema,
                    predicate: Option<Predicate>,
                    fs: FileSystem): Observable<Row> {

    val reader = ParquetReaderFns.createReader(path, predicate, Option(projectionSchema))
    return Observable.create<Row> { subscriber ->
      subscriber.onStart()
      ParquetRowIterator(reader).forEach {
        subscriber.onNext(it)
      }
      subscriber.onCompleted()
    }
  }

  override fun writer(schema: Schema,
                      path: Path,
                      fs: FileSystem): HiveWriter = object : HiveWriter {
    init {
      ParquetLogMute()
    }

    // hive is case insensitive so we must lower case the fields everything to keep it consistent
    val avroSchema = AvroSchemaFns.toAvroSchema(schema, caseSensitive = false)
    val writer = ParquetRowWriter(path, avroSchema, fs)
    val serializer = AvroRecordSerializer(avroSchema)

    override fun write(row: Row) {
      val record = serializer.toRecord(row)
      writer.write(record)
    }

    override fun close() {
      logger.debug("Closing dialect writer")
      writer.close()
    }
  }
}